Add community detection

rahuljoglekar47 opened the pull request:
## Fixes # .
Code should be reasonably efficient (hasn't been profiled yet) as a result of restructuring it; the Python logging module was also implemented

## Summary/Motivation:
Take in a Pyomo optimization model into function detect_communities, and then return a map of the communities within the model (based on various user-specified parameters such as the node type, whether the graph is weighted, etc.)

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

jsiirola committed:
Updating job naming

jsiirola committed:
Replacing github actions with unified workflows

jsiirola committed:
Rename github workflows to match codecov detection rules

jsiirola committed:
Jenkins: do not duplicate the source= coverage directive

jsiirola committed:
Rename github workflows to match codecov detection rules

jsiirola committed:
Reverting changes to GitHub CI name

jsiirola committed:
Add wheel to python environment

jsiirola committed:
Distribute branch testing across python versions

jsiirola committed:
Fixing matrix definition for branch workflow

jsiirola committed:
NFC: converting tab whitespace to space

jsiirola committed:
Updating readme

jsiirola committed:
Fixing #defines for MA27/57 __declspec directives

jsiirola committed:
Adding DL dependency (for linux/osx)

jsiirola committed:
Merge remote-tracking branch 'robby/hsl_interface' into hsl_interface

michaelbynum committed:
minor updates to gdp and gdpopt

emma58 committed:
Adding an option to relax_integer_vars so that by default it descends into deactivated components, but you can ask it not to.

mrmundt committed:
Trying different PyUtilib branch:

emma58 committed:
Merge branch 'master' into relax-discrete-var-fix

Robbybp committed:
Addressing commments

jsiirola committed:
help_solvers: add 10-second timeout for NEOS connection

jsiirola committed:
Merge branch 'master' into unified-workflow

emma58 committed:
Adding a default filtering that checks if a constraint generated by FME is implied by the variable bounds before it adds it to the model.

ZedongPeng committed:
fix if-else for boolean configurations

emma58 committed:
Adding a tests with filtering for the chull test as well, to make sure we don't lose anything important

emma58 committed:
Changing argument name

blnicho committed:
Fixing typos

ZedongPeng committed:
Merge branch 'master' into new-lp-nlp

jsiirola committed:
Merge pull request #1426 from jsiirola/unified-workflow

Unified GitHub Actions workflow

qtothec committed:
Merge pull request #1391 from ZedongPeng/new-lp-nlp

mindtpy: implementation of lp/nlp algorithm, bug fix and documentation update

Robbybp committed:
class methods for accessing logger

Robbybp committed:
remove references to IP method

Robbybp committed:
method for reallocation from IP method

Robbybp committed:
typos

Commit:
:hammer: Use `store_in_cplex()` instead of ctxmanager

- Calling a method to "finalise" the data objects is more explicit than `__exit__()` and doesn't rely on `nullcontext()` from CPython

Commit:
:books: Formatting

michaelbynum committed:
Merge pull request #1419 from Robbybp/hsl_interface

PyNumero: HSL interface

michaelbynum committed:
Merge branch 'master' into gdp_updates

mrmundt committed:
Merging upstream

emma58 committed:
Adding a post-processing method to FME so that you can eliminate redundant constraints by just going through and checking which are implied by the others.

Robbybp committed:
Use increase_memory_allocation rather than set_ and get_memory_allocation

Robbybp committed:
Update test for reallocation

michaelbynum committed:
Merge pull request #1432 from michaelbynum/gdp_updates

Gdp updates

Robbybp committed:
adjust log format

Robbybp committed:
simple context for logging linear solve info

emma58 committed:
Fixing BigM Suffixes so that we do look on child blocks of Disjuncts, but only later when we are transforming the constraints

emma58 committed:
Adding name buffer arg to _warn_for_active_disjunct

emma58 committed:
Only passing indexed components through the _transform_disjunct and _transform_disjunction methods, SimpleComponents can go straight to transfrom_somethingData

Robbybp committed:
Remove get_infog from IP method

Robbybp committed:
Update test to use synthetic example

emma58 committed:
Updates post-set-rewrite, PEP-8ing, switching order of original index and our index in disaggregation constraints (which is why the baselines changed)

emma58 committed:
merging master

michaelbynum committed:
Merge pull request #4 from Robbybp/pynumero_interior_point

solver class, memory reallocation, and regularization

michaelbynum committed:
Merge branch 'master' into pynumero_interior_point

emma58 committed:
Skipping tests for post process if solver not available

emma58 committed:
Reraising exceptions from the methods to get at the mappings

emma58 committed:
Fixing a typo

jsiirola committed:
Fix component_data_objects for scalar components with no len()

Fixes #1435.

qtothec committed:
adding "int" to the typehint

jsiirola committed:
Ensure block rules are always called

emma58 committed:
Making how we treat fixed variables in chull an option

jsiirola committed:
Update pyomo.dae now that Blocks always fire their own rules

qtothec committed:
small correction

jsiirola committed:
Merge pull request #1428 from emma58/relax-discrete-var-fix

Fix handling of deactivated blocks in relax_integer_vars transformation

emma58 committed:
Adding tighter M estimation when user promises that fixed vars will remain so

jsiirola committed:
Merge pull request #1437 from qtothec/rangeset_typehint_docfix

[NFC] RangeSet docstring type hint

emma58 committed:
Removing debugging

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

michaelbynum committed:
interior_point: utilizing a linear solver results object

michaelbynum committed:
a little pynumero reorganization

michaelbynum committed:
addressing issue with mumps interface

michaelbynum committed:
reorganizing pynumero slightly

michaelbynum committed:
Merge branch 'pynumero_linalg' into pynumero_interior_point

michaelbynum committed:
updating tests and imports

michaelbynum committed:
Merge branch 'pynumero_linalg' into pynumero_interior_point

jsiirola committed:
Do not 'densify' indexed blocks with no rule

jsiirola committed:
Update Complementarity component

- track changes in Block's handling of rules
- improve consistency handling Complementarity.Skip in set_value
- incorporate Initializer and disable_methods from core.base.util

michaelbynum committed:
updating imports

michaelbynum committed:
updating tests

michaelbynum committed:
Merge branch 'pynumero_linalg' into pynumero_interior_point

renkekuhlmann committed:
using gams gdx instead of put files to read result

blnicho committed:
Fixing typos

renkekuhlmann committed:
added gdxDataReadDone

renkekuhlmann committed:
improve gams call

jsiirola committed:
Merge pull request #1436 from jsiirola/comp-iter-fix

Fix component_data_objects for scalar components with no len()

mrmundt committed:
Merge branch 'master' of https://github.com/mrmundt/pyomo

emma58 committed:
Merge branch 'master' into chull-rewrite

emma58 committed:
Merge branch 'master' into fme-improvements

emma58 committed:
Updating FME post-process now that relax integrality works

emma58 committed:
Removing no-longer-needed import

michaelbynum committed:
interior point cleanup (mostly uniform linear solver API)

michaelbynum committed:
interior point cleanup

michaelbynum committed:
interior point cleanup

michaelbynum committed:
interior point cleanup

michaelbynum committed:
interior point cleanup

michaelbynum committed:
interior point cleanup

michaelbynum committed:
working on interior point timing

michaelbynum committed:
Interior Point: simplifying evaluations

michaelbynum committed:
simplifying evals

renkekuhlmann committed:
added option to disable keepfile output

michaelbynum committed:
interior point cleanup

michaelbynum committed:
pynumero linalg cleanup

michaelbynum committed:
Merge branch 'pynumero_linalg' into ip_evals

michaelbynum committed:
interior point cleanup

michaelbynum committed:
Merge pull request #3 from Pyomo/master

merge master into pynumero_interior_point

michaelbynum committed:
Merge pull request #6 from Pyomo/master

merge master into pynumero_interior_point

michaelbynum committed:
Merge remote-tracking branch 'robbybp/pynumero_interior_point' into pynumero_interior_point

blnicho committed:
Merge pull request #1410 from Robbybp/dae-tools

Solving for consistent initial conditions

blnicho committed:
Fixing typo

renkekuhlmann committed:
consistent gams job name among interfaces

renkekuhlmann committed:
read gdx solution only if available

renkekuhlmann committed:
add  to gams_writer test baselines

renkekuhlmann committed:
fixed keepfiles output clause

renkekuhlmann committed:
added more offlistings to test baselines

renkekuhlmann committed:
return None instead of NaN if no solution returned

michaelbynum committed:
interior point: some performance improvements

renkekuhlmann committed:
fixed gdx solution read to address constant (thus removed) equations

DLWoodruff committed:
Updating tests for inverse reduced hessian

michaelbynum committed:
interior point: updating tests

renkekuhlmann committed:
fixed gdx read of solution attributes for python 2.7

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

michaelbynum committed:
interior point: adding a method to load variable values back into the pyomo model

mrmundt committed:
Changing badge to GA

mrmundt committed:
Fixing badge path

mrmundt committed:
Trying without the label

michaelbynum committed:
interior point: updating imports

mrmundt committed:
Removing appveyor YML file

michaelbynum committed:
interior point: updating tests

michaelbynum committed:
Merge pull request #1439 from michaelbynum/pynumero_linalg

Pynumero Linalg Updates

michaelbynum committed:
Merge branch 'master' into pynumero_interior_point

mrmundt committed:
Moving around some README stuff

jsiirola committed:
Prevent codecov notification until all builds complete

jsiirola committed:
Include coverage report in Travis build log

michaelbynum committed:
Merge pull request #1438 from jsiirola/block-rule-init

Ensure block rules are always called

michaelbynum committed:
Merge pull request #1360 from michaelbynum/fbbt

FBBT Numerical Issues

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

mrmundt committed:
Adding FICO Xpress via pip/conda installs to GA Workflows

rahuljoglekar47 committed:
Setting up framework to add community detection

jsiirola committed:
Merge pull request #1447 from mrmundt/remove_appveyor

Remove Appveyor testing

jsiirola committed:
Merge pull request #1451 from jsiirola/travis-codecov-updates

Coverage configuration updates

jsiirola committed:
Prevent exception for transforamtions missing doc

Transformations that fail to define a doc string should not cause `pyomo
help --transformations` to generate an exception.  This also adds a basic test
that help_transformations() runs and generates reasonable output.

jsiirola committed:
Fix typos in .codecov.yml

The 'notify:' section needs to by under 'codecov:' and not 'coverage:'

jsiirola committed:
Merge pull request #1452 from mrmundt/xpress

Add Xpress to Github Actions

jsiirola committed:
Merge pull request #1454 from jsiirola/fix-help-transformations

Prevent exception for transformations missing doc string

blnicho committed:
Merge pull request #1455 from jsiirola/codecov-config

Fix typos in .codecov.yml

rahuljoglekar47 committed:
Delete detection.py

rahuljoglekar47 committed:
Delete profile_code.py

rahuljoglekar47 committed:
Delete test_efficiency.py

rahuljoglekar47 committed:
Setting up framework to add community detection

rahuljoglekar47 committed:
Merge remote-tracking branch 'origin/add_community_detection' into add_community_detection

# Conflicts:
#	pyomo/contrib/community_detection/detection.py

jsiirola committed:
Disable travis use of S3 codecov uploader

jsiirola committed:
Use @wraps so GDP transformation methods inherit docstrings

jsiirola committed:
Quote component names in exception/logger messsages

...also remove trailing whitespace

rahuljoglekar47 committed:
Changed variable names to make more sense

rahuljoglekar47 committed:
Added some helpful comments and changed variable names to make more sense

jsiirola committed:
Make while look compatible with OSX

jsiirola committed:
Make codecov.sh exit with 1 on error

jsiirola committed:
Increase codecov upload retries to 4 with londer delay

jsiirola committed:
Add codecov upload retries to travis builds

jsiirola committed:
Reduce the number of reports before publishing coverage data

renkekuhlmann committed:
removed disabling keepfiles output using tee

renkekuhlmann committed:
make improved gams call default but not overwriting

renkekuhlmann committed:
use attempt_import for gdxcc

emma58 committed:
Removing domain for constraint filtering callback and adding what is hopefully a slightly more helpful error message if it turns out not to be what we expect.

emma58 committed:
Fixing typos

jsiirola committed:
Merge pull request #1421 from emma58/chull-rewrite

Rewrite of the GDP Chull transformation

jsiirola committed:
Merge pull request #1457 from jsiirola/update-travis-codecov

Coverage reporting updates

emma58 committed:
Adds doc attribute to FME transformation, but now it has it in two places?

emma58 committed:
Merging chull changes into FME tests (again changes order of transformed constraints, but should be more robust now)

rahuljoglekar47 committed:
Added some helpful comments and changed variable names to make more sense

rahuljoglekar47 committed:
Added one line to the docstring for _generate_model_graph

qtothec committed:
revert implementation of rounding, which may lead to improper exclusion of feasible solution points

qtothec committed:
add deprecation warning

rahuljoglekar47 committed:
Modified community_graph slightly to account for the possibility of a model with multiple objective functions (and reformatted code)

jsiirola committed:
Merge pull request #1458 from qtothec/fix_pyomo_constraint_tightener

Fix and deprecate constraint tightener transformation

rahuljoglekar47 committed:
Added an import statement (from pyomo.core import sum_product) so that the LP_unique_duals model could be used without running into an error

emma58 committed:
Fixes double-registering of FME transformation (which fixes #1453), and removes import from __init__.py

emma58 committed:
Making post-processing tolerance an argument to the function

rahuljoglekar47 committed:
Cleaned up test functions and added more test cases; however, there is a problem with test_communities_2 that I am currently working on. For some reason, even with a seed number for the Louvain community detection, the community map generated for QCP_simple is not always the same.

emma58 committed:
Reducing some calls to value by making sure we do what we can when we put the constraints in dictionary form

rahuljoglekar47 committed:
Reformatted test file and added title docstring (Community Detection Test File)

jsiirola committed:
Fix scalar block construction if self was not already in _data

This works around an issue encountered where rules were called twice for
derived scalar blocks, where the derives scalar class did not initialize
`_data[None] = self`.

michaelbynum committed:
Merge branch 'master' into pynumero_interior_point

michaelbynum committed:
interior point: updating imports

jsiirola committed:
Adding test to verify construction for "concrete-only" blocks

michaelbynum committed:
flushing out abstract methods for interior point interface

jsiirola committed:
Merge pull request #1440 from emma58/fme-improvements

FME transformation constraint pruning improvements

michaelbynum committed:
adding some pyomo-based method to the interior point interface

rahuljoglekar47 committed:
Changed a test case (changed QCP_simple to QP_simple because I could not resolve the issues I found with QCP_simple)

rahuljoglekar47 committed:
Modified titles of files - also, this is the last commit before logger functionality is added

jsiirola committed:
Fall back on GAMS "put" interface if gdx is unavailable

jsiirola committed:
Fixing typo

jsiirola committed:
Updating GAMS tests

jsiirola committed:
Merge pull request #1459 from jsiirola/scalar-block-construct

Fix construction of scalar derived blocks

renkekuhlmann committed:
Merge pull request #1 from jsiirola/gams_gdx

Support PUT and GDX GAMS interfaces

rahuljoglekar47 committed:
Removed else case for checking node_type in community_graph.py because this will be handled in detection.py by the event logger; also added log_level as an argument to detect_communities and added some basic uses of the logging module

Commit:
:hammer: Refactor quadratic objective handling

michaelbynum committed:
Merge pull request #1450 from michaelbynum/pynumero_interior_point

Interior Point

jsiirola committed:
Merge pull request #1446 from renkekuhlmann/gams_gdx

Add GDX interface for returning solution values from GAMS

Commit:
:bug: Cast quadratic coefficients to floats explicitly

rahuljoglekar47 committed:
Added function (check_for_correct_arguments) that uses the logger and determines whether arguments given to detect_communities are correct

rahuljoglekar47 committed:
Forgot to add conditionals for default arguments of type None for file_destination and random_seed; thus the last commit failed all test cases because None was returned instead of the community maps

jsiirola committed:
Merge pull request #1416 from flexciton/perf/cplex_direct

CPLEXDirect performance improvements

rahuljoglekar47 committed:
The following things were done in this commmit (all relevant to logging):
- _event_log function was created (this is called in _generate_model_graph) to record information about the model and the resulting model_graph
- If the file path given to detect_communities does not exist then an error will be logged along with the file path created by os.makedirs (in the _write_to_file function)
- In detect_communities, important information about the decomposition/number of communities detected will be logged

rahuljoglekar47 committed:
1) Added a check in the detect_communities logger function to make sure the model is an instance of ConcreteModel
2) Added a test case that checks to make sure None is returned when given bad arguments (as well as an empty dictionary when given an empty model)
3) Removed a redundant import statement from test_detection.py

jsiirola committed:
Remove an unneeded 'tee' of the codecov output

blnicho committed:
Merge pull request #1462 from jsiirola/gha-codecov-retry

Remove an unneeded 'tee' of the codecov output

rahuljoglekar47 committed:
Added comment about use of check_for_correct_arguments function

rahuljoglekar47 committed:
Changed some of the docstring for the detect_communities function

rahuljoglekar47 commented:
Please let me know if there is a better way to determine the variables in a given constraint equation (as I did in lines 57 and 58 of community_graph.py); I do something like this (adding the string of a variable to a list) many times in this file so I was wondering if this is the best way to create a list of variables in a given equation (or constraints that contain a given variable).

bernalde commented on pyomo/contrib/community_detection/detection.py:
```diff
@@ -0,0 +1,156 @@
+"""Community Detection Code - Rahul Joglekar"""
```
Usually contributed files to Pyomo do not include the author.

bernalde commented on pyomo/contrib/community_detection/detection.py:
```diff
@@ -0,0 +1,156 @@
+    return community_map
+
+
+def check_for_correct_arguments(model, node_type, with_objective, weighted_graph, file_destination, log_level,
```
Since this is some sort of internal function with a generic name I would suggest renaming it as `_check_for_correct_arguments`

bernalde commented on pyomo/contrib/community_detection/tests/test_detection.py:
```diff
@@ -0,0 +1,321 @@
+
+
+class TestDecomposition(unittest.TestCase):
+
```
define a variable `random_seed=5` here and use it below instead of hardcoding the 5 in every instance

bernalde commented on pyomo/contrib/community_detection/tests/test_detection.py:
```diff
@@ -0,0 +1,321 @@
+    #      Total    const       NL      DLL
+    #         37       35        2        0
+    #
+    #  Reformulation has removed 1 variable and 1 equation
```
The comments here and above can be removed. You can add a single line saying that it comes from instance XXX.gms in minlplib.com

bernalde commented on pyomo/contrib/community_detection/tests/test_detection.py:
```diff
@@ -0,0 +1,321 @@
+    model = m = ConcreteModel()
+    m.i1 = Var(within=Integers, bounds=(0, 1E15), initialize=0)
+    m.i2 = Var(within=Integers, bounds=(0, 1E15), initialize=0)
+    m.i3 = Var(within=Integers, bounds=(0, 1E15), initialize=0)
```
The upper bounds of integer variables being this large seem wrong. Just set 100 (GAMS does that by default).

bernalde commented on pyomo/contrib/pynumero/examples/structured/tests/test_nlp_transformations.py:
```diff
@@ -12,7 +12,8 @@
 import os
 
-from .. import numpy_available, scipy_available
+#from .. import numpy_available, scipy_available
```
This shouldn't be here. These changes are not part of this PR, so revert them to original state.

bernalde commented on pyomo/solvers/tests/models/LP_unique_duals.py:
```diff
@@ -8,7 +8,8 @@
 
-from pyomo.core import ConcreteModel, Param, Var, Expression, Objective, Constraint, RangeSet, ConstraintList, NonNegativeReals, Suffix, summation
+from pyomo.core import ConcreteModel, Param, Var, Expression, Objective, Constraint, RangeSet, ConstraintList, \
+    NonNegativeReals, Suffix, summation, sum_product
```
Same case here, do not edit any of the files that do not belong to the original contribution.
Were these changes required for the tests of communities to work?

bernalde commented on pyomo/contrib/community_detection/detection.py:
```diff
@@ -0,0 +1,156 @@
+"""Community Detection Code - Rahul Joglekar"""
+from pyomo.core import ConcreteModel
+from contrib.community_detection import community_graph
+import community
```
This is a point that we haven't discussed yet but that seems important. What happens if the user does not have the package `community` installed? We should have a way of checking this import and besides hopefully having this package installed along with Pyomo.
Follow the code here https://github.com/Pyomo/pyomo/blob/07d3cdbde1139d3f926d322132c1995e00187b62/pyomo/contrib/mindtpy/mip_solve.py#L28 to attempt an import of the package and add the documentation of the package in the `communities.rst` file

bernalde reviewed (changes requested):
Hi Rahul, this looks fine and it should be working.
My two main requests are:

1. Adding an attempt import for the community package. If the user does not have it installed, then you should also skip the tests (at least those tests that require you running the community detection algorithm).
2. Adding the documentation of your contributed package as an .rst file

ghackebeil committed:
PySP is divorcing pyutilib.enum (#1464)

* use builtin enum for piecewise

* misc fix for set rewrite behavior

* nfc: whitespace cleanup

* use builtin enum

* misc import fix that caused local test failure

* use something enum-like for the _EnumValueWithData replacement

qtothec committed:
revert changes to unrelated files

qtothec committed:
Merge remote-tracking branch 'upstream/master' into joglekar_community_detection

qtothec committed:
code for pyomo integration and suggestions of changes

qtothec committed:
correctly skip tests if packages not available

qtothec committed:
adding stub for documentation

rahuljoglekar47 committed:
Merge pull request #3 from qtothec/joglekar_community_detection

Suggested edits

rahuljoglekar47 committed:
Removed author name from title docstrings; reverted (fixed?) a default argument or detect_communities from the merge with Qi's branch; changed comment for create_model_5() in test_detection.py

rahuljoglekar47 committed:
Fixing hardcoded random_seed value (last of David's comments)

rahuljoglekar47 committed:
Reformatted file

rahuljoglekar47 committed:
Added assert statements for all of the arguments given to detect_communities, and updated test function to test for the correct responses to bad arguments and to test for the correct logging errors/warnings

rahuljoglekar47 committed:
Reformatted and modified comments

rahuljoglekar47 committed:
Last commit before changing structure of community_graph (also did some minor reformatting)

rahuljoglekar47 committed:
Changed structure of community_graph.py to use the id() of model elements instead of their str() values, making many parts of the code much simpler. Note that this code fails 4 of the 7 test cases due to the lack of sorting the edges that are stored in edge_weight_dict. That will be implemented next.

rahuljoglekar47 committed:
\Handled the issue with edge weights being added incorrectly due to ordering issues - note: this commit does not sort the lists in community_map but when they are sorted, all test cases pass except for test_communities_3(). This commit thus fails several of the test cases (somewhat inconsistently).

rahuljoglekar47 committed:
(Intermediate commit) community_graph.py was changed by adding a few lines to ensure that the networkx graph that is returned results in consistent community detection; two test cases were swapped (because of an issue with my IDE running independent tests)

rahuljoglekar47 committed:
Consistent community detection now! The networkx graph is now constructed at the very end (which allows for consistent ordering of nodes and edges). However, this commit is still messy - comments have not been updated, lines 158-162 need to be made more efficient, etc.; as for test_detection.py one test case had to be modified now that the specific ordering has changed but all test cases are passed with this commit

rahuljoglekar47 committed:
Deleted one unused import statement

rahuljoglekar47 committed:
Design of community_graph.py was changed again due to issues with using ComponentSet and ComponentMap (listed below). Instead, the id() function was used to store Var/Constraint/Objective in edge_weight_dict and edge_set and an id_map was created so that the strings of the Var/Constraint/Objective could still be put into the model_graph (networkx graph). All test cases pass with this commit.

Issues:
- ComponentMap does not group identical keys, leading to issues with using the get method for Var/Constraint/Objective
- neither ComponentMap nor ComponentSet are useful for determining membership of Var/Constraint/Objective

rahuljoglekar47 committed:
str_map is needed in detect_communities.py, so str() was used instead of id() (only a moderate difference in terms of memory usage, based on sys.getsizeof() function)

rahuljoglekar47 committed:
In detection.py, string_community_map (which is what we used to refer to simply as community_map) is now used to create community_map, a dictionary mapping the community number (integers) to lists of actual Pyomo variables/constraints/objectives. This community_map is what is returned. The test function has not yet been updated, so all test cases fail, but the two-line conversion seems to be working properly. Minor changes made to community_graph.py (formatting issues and added the string_map to the return statement)

rahuljoglekar47 committed:
Added tests for a community_map that maps communities to lists of Pyomo variables/constraints/objectives. Still have not implemented an option for the user to select between this community_map as opposed to one made up of strings, so all of the 6 test functions that end in "str" (these are the old test functions) fail.

rahuljoglekar47 committed:
Added string_output as an argument for detect_communities and consolidated test functions

rahuljoglekar47 committed:
Added constraint_variable_map to community_graph.py and detection.py; next commit will use this map in detection.py

rahuljoglekar47 committed:
Implemented usage of constraint_variable_map in detection.py; Changed community_graph.py to construct the string_map within the node_type conditionals; also went back to adding nodes one by one as opposed to from a very large list

rahuljoglekar47 committed:
Updated test cases to reflect the new structure of the community map and removed the function calls where string_output=True from all except one test function. All test cases pass with this commit.

rahuljoglekar47 committed:
Added a helper function for test_detection.py that saves about 150 lines thanks to a lot of duplicate code

rahuljoglekar47 committed:
Reformatted test_communities_1

rahuljoglekar47 committed:
Moved around node type conditionals in community_graph.py; reformatted test_detection.py (last commit before implementing bipartite graphs)

rahuljoglekar47 committed:
Implemented node type 'b' as an option for community detection (bipartite graph) - comments still need to be updated; also simplified edge_weight_dict and edge_set by sorting tuples as they are added

rahuljoglekar47 committed:
Wrote get_edge_list() and get_adj_list() functions in detection.py; these functions (as well as all others) need comments and docstrings updated

rahuljoglekar47 committed:
Updated all comments and dosctrings; wrote get_edge_list and get_adj_list in detection.py (also deleted _write_to_file in community_graph.py)

rahuljoglekar47 committed:
Modified an import statement in community_graph.py (which will hopefully fix the CI checks)

rahuljoglekar47 committed:
Removed "weighted" bipartite graph because all edges were recorded as weight=1; Modified import statements in test_detection.py and detection.py

rahuljoglekar47 committed:
Add node for all objective function (would be missed in case of disconnected graphs)

rahuljoglekar47 committed:
Changed community_graph.py to collapse bipartite graphs into variable and constraint graphs (generally around 75% faster for constraint node graphs, slightly faster for variable node graphs, and about the same speed for bipartite graphs)

rahuljoglekar47 committed:
Modified test_detection.py only due to ordering changes in the community maps from changes in community_graph.py

rahuljoglekar47 committed:
Mainly added comments for community_graph.py; changed order of tuple in test function; minor comment changes to detection.py

rahuljoglekar47 committed:
Changed community map returned for node_type 'v' to have constraint list given first; changed test_detection.py to reflect this change

rahuljoglekar47 committed:
Changed the third item returned by community_graph.py depending on node_type; also modified test_detection.py because test_communities_4 and test_decogo_2 failed unexpectedly

rahuljoglekar47 committed:
NOTE: All commits from the one on June 3 titled "Implemented node type 'b'..." (875c9c4fd0e2093881a24f9a472fb3c7b69d0f2b) up until the third to last commit on June 8 titled "Changed community map returned..." (13bbe833ce863540cc67f28935e95a67c951b7a7)  has an INCORRECT community map when node_type='b'. The line in detection.py "if isinstance(node, Var)" never gets executed, even when the node is actually a Var. This was fixed with variable_node_set, which has been implemented since the second to last commit on June 8 titled "Changed the third item..." (8ea44af5c078d226755d8db4f58f6b6494b4a25a). (This commit is only for the sake of documenting this error; but also some import statements were adjusted/reformatted)

rahuljoglekar47 committed:
Added graph drawing capability - not finished yet
